src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -0,0 +1,448 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions
+{
+    using static EmbeddedSyntaxHelpers;
+    using static RegexHelpers;
+
+    using RegexToken = EmbeddedSyntaxToken<RegexKind>;
+    using RegexTrivia = EmbeddedSyntaxTrivia<RegexKind>;
+
+    /// <summary>
+    /// Produces tokens from the sequence of <see cref="VirtualChar"/> characters.  Unlike the native C# and VB
+    /// lexer, this lexer is much more tightly controlled by the parser.  For example, while C# can have trivia
+    /// on virtual every token, the same is not true for RegexTokens.  As such, instead of automatically lexing
+    /// out tokens to make them available for the parser, the parser asks for each token as necessary passing
+    /// the right information to indicate which types and shapes of tokens are allowed.
+    /// 
+    /// The tight coupling means that the parser is allowed direct control of the position of the lexer.
+    /// 
+    /// Note: most of the time, tokens returned are just a single character long, including for long sequences
+    /// of text characters (like ```"goo"```).  This is just three <see cref="RegexTextNode"/>s in a row (each
+    /// containing a <see cref="RegexKind.TextToken"/> a single character long).
+    /// </summary>
+    internal struct RegexLexer
+    {
+        public readonly ImmutableArray<VirtualChar> Text;
+        public int Position;
+
+        public RegexLexer(ImmutableArray<VirtualChar> text) : this()
```

Comment:
What is `: this()` for?

Reply:
Should the lexer hold onto the `RegexOptions` (so it could be removed as parameter to various APIs), or can the options change from one call to the next?

Reply:
The options can change from call to call.  Users can change regex options in-situ in the pattern itself.  i.e. ```(?insmx)```